### PR TITLE
Allow optional -D/--pgdata flag for PG upgrade

### PR DESCRIPTION
**Pull request by stevenharman**
Postgres can be installed with different data directories, or with a non-default location. This allows a user to upgrade those instances too.

- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/master/CONTRIBUTING.md)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Does your build pass `brew audit --strict <formula>` (after doing `brew install <formula>`)?

-----

This is for an external command, rather than a formula, so I'm not sure the later three apply. That said, I have tested this locally with both the defaults and a `--pgdata ~/.pgdata` option and successfully upgraded from PG 11 to 12.

### CODE REVIEW

File: cmd/brew-postgresql-upgrade-database.rb [hunk 1]
```diff
@@ -1,12 +1,24 @@
...
+#:  * `postgresql-upgrade-database` [options]
 #:    Upgrades the database for the `postgresql` formula.
+#:
+#:    -D, --pgdata    Location of the database configuration
+#:                    files. Same as the pg_ctl -D flag.
+
+require "optparse"
+options = {}
+OptionParser.new do |opts|
+  opts.on("-D", "--pgdata DATADIR", "Location of the database configuration files.") do |data_dir|
```
Comment (stevenharman) [inline cmd/brew-postgresql-upgrade-database.rb]: I mirrored the option names from `pg_ctl` in hopes of keeping it consistent for folks using that option. But I'm happy to use another name.

**Comment by MikeMcQuaid** [comment]
> Postgres can be installed with different data directories, or with a non-default location. This allows a user to upgrade those instances too.

That's never done by Homebrew's formula, though, right?

**Comment by stevenharman** [comment]
> That's never done by Homebrew's formula, though, right?

I think we're talking about different things here, and it's probably my fault for saying "can be installed into different data directories."

Postgres is installed, by Homebrew, into wherever Homebrew is configured to install things. On my machine, that's:

```bash
brew --prefix
/usr/local
```

That's where the PostgreSQL binaries and such live. So when you `pg_ctl`, `psql`, etc... that's what's running. Those binaries use the configuration and backing data files, as defined by the `DATADIR` to spin up a running Postgres instance. i.e., a running Postgres database instance is the binaries _plus_ the config + data files. (See also: [PGDATA docs](https://www.postgresql.org/docs/current/storage-file-layout.html))

When initializing, starting, or stopping an instance of Postgres, you can tell it which `DATADIR` to use via the `-D` or `--pgdata` flags. 

```bash
initdb -D <some/path>

pg_ctl -D <some/path> start

pg_ctl -D <some/path> stop
```

If you don't provide this option, it uses the `PGDATA` env var, which has a default value of something like `/usr/local/var/postgres/` on my machine. But that's just a default, and so it's also what things like `brew services start postgresql` will use, _as a default_. `brew info` mentions this, though it's not very obvious:

```bash
$ brew info postgresql
postgresql: stable 12.1 (bottled), HEAD
[... snip ...]
==> Caveats
To migrate existing data from a previous major version of PostgreSQL run:
  brew postgresql-upgrade-database

To have launchd start postgresql now and restart at login:
  brew services start postgresql
Or, if you don't want/need a background service you can just run:
  pg_ctl -D /usr/local/var/postgres start
[... snip ...]
```

That last line ⬆️ shows the option, but it's using the default value, making this all the more confusing. The Postgres upgrade script was relying on that default always being used, and thus it was hard-coded in. But it is totally possible to have a single Hombrew-installed PostgreSQL, and run multiple instances of it, each pointing at a different data directory. Or to do what I do, and only run a single instance, but keep the data w/in my user space.

Does that make sense?

**Comment by MikeMcQuaid** [comment]
> Does that make sense?

Yes, thanks for the complete explanation.

My thinking is that unless `brew install postgresql` or `brew postinstall postgresql` installs into another directory I'm hesitant to add functionality to this script (that will be indefinitely supported/relied on) that exists only to operate on configurations that Homebrew hasn't set up.

**Comment by stevenharman** [comment]
I understand the hesitation, for sure. However, the option to specify a datadir is so fundamental to PostgreSQL that it's not really option. `initdb`, which is used by `brew post install postgresql`, requires either the `PGDATA` Env var be set, or the `-D` flag to be passed. It could be argued that the Homebrew Postgres formula has a bug; it ignores the `PGDATA` Env var, and doesn't accept a `-D` flag, instead forcing a hard-coded path upon users.

Similarly, the `pg_ctl` command requires either `PGDATA` or the `-D` flag, and that flag is again hard-coded for `brew services` via the formula's `plist_options` (https://github.com/Homebrew/homebrew-core/blob/69150f8c91272c0df35618df9eab7377f0943ee2/Formula/postgresql.rb#L88).

In its current form, `brew postinstall postgresql`  is creating an Postgres database cluster, which a user might never use. And an unfortunate side effect of the same hard-coding in the database upgrade script is that other clusters cannot be upgraded with this script.

My hope was to allow an optional `datadir` to be specified in the upgrade, while still defaulting to the same hard-coded values that the formula itself uses. And I realize now, the upgrade should probably also respect any `PGDATA` Env var that's set.

Does that make a better case for the proposed change? Or would a corresponding change in the formula itself need to be proposed and accepted? 

Thank you for your time and all of the effort you put into this project.

**Comment by MikeMcQuaid** [comment]
> It could be argued that the Homebrew Postgres formula has a bug; it ignores the `PGDATA` Env var, and doesn't accept a `-D` flag, instead forcing a hard-coded path upon users.

This is intentional behaviour.

> My hope was to allow an optional `datadir` to be specified in the upgrade, while still defaulting to the same hard-coded values that the formula itself uses. And I realize now, the upgrade should probably also respect any `PGDATA` Env var that's set.

Homebrew intentionally filters out all environment variables to make a clean and consistent configuration. 

As a result and discussion above: sorry but passing on this PR.

> Thank you for your time and all of the effort you put into this project.

Thanks for the kind words and the PR :heart:

**Pull request closed without merging**